### ISSUE TITLE
De-dupe notifications already recieved on the desktop for equivilent app.

### ISSUE THREAD

**Issue by stuaxo** [GSConnect/gnome-shell-extension-gsconnect#1070]
Run something like Microsoft Teams on your phone and on your desktop (or whatsapp on phone + whatsdesk on desktop).

Get some notifications from the desktop app.

5 minutes later get all the same from the mobile app.

I'd really like some de-duping here, getting everything twice (or sometimes more !) is so distracting.

**Comment by not-a-dev-stein** [comment]
For me it's Telegram, same experience here, don't know if it's possible but I'd really like to see something like this.

**Comment by sonnyp** [comment]
I don't think this is possible without significant efforts so perhaps not that good of a "good first issue".

You can disable notifications for specific applications in KDE Connect so you may disable sending Microsoft Teams or WhatsApp notifications to your computer.

Looks like dunst might also be able to do that https://github.com/dunst-project/dunst/issues/571 - if anyone can confirm - it'd be great to have a HOWTO on the wiki.

**Comment by stuaxo** [comment]
Doing this in dunst sounds like a good idea for a proof of concept.

I had a look on my phone, I already disable notifications on the phone, but that doesn't stop them getting KDE Connect sending them.

You're right this probably isn't a "first issue", 

I guess "Notify" https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/80c3d156fbc61c97d3a3a004172f41056d8ff316/src/service/components/notification.js#L369

would need to have a list of apps to suppress notifications from.

The other side would probably take most of the work -
You'd need a per app menu, that had "Suppress notifications when" and then you could add "App is running..." where you could choose the app.

When the app was running it would be added to the suppress notifications list, and removed when it stops.

**Comment by ferdnyc** [comment]
I don't know why this is marked "good first issue", I've removed that.

The problem with de-duplication (well, one problem) is that GSConnect isn't necessarily supposed to have access to notifications on your desktop that it **didn't** create. It probably _does_, today, but there's a good chance that eventually gets locked down so it doesn't, anymore.

> The other side would probably take most of the work -
> You'd need a per app menu, that had "Suppress notifications when" and then you could add "App is running..." where you could choose the app.

The thing is, with push notifications, there's nothing to say that an app is actually _running_ even if notifications are coming through.

Both sides already have the ability to turn off mirroring of notifications on an app-by-app basis, and that configuration already applies per paired device. So, it seems like the simplest solution might be, if you know that your phone and laptopA both get notifications from appB, just disable appB notification mirroring to the other device, from both ends.

**Comment by jnnkB** [comment]
> Both sides already have the ability to turn off mirroring of notifications on an app-by-app basis, and that configuration already applies per paired device. So, it seems like the simplest solution might be, if you know that your phone and laptopA both get notifications from appB, just disable appB notification mirroring to the other device, from both ends.

While this is the easiest solution, it's not the most convenient: If you only sometimes run appB on your laptop and also your laptop and phone aren't always connected, de-duping notifications would ensure that you are notified as often as possible.

**Comment by stuaxo** [comment]
Sounds like there needs to be a centralised service for notifications that handles the de-duping, that isn't part of GS-Connect.

On the other hand, it could be an offshoot of this project, as this is the main scenario where you are going to get duplicates like that.

**Comment by ferdnyc** [comment]
Yeah, this...

> GSConnect isn't necessarily supposed to have access to notifications on your desktop that it **didn't** create. It probably _does_, today, but there's a good chance that eventually gets locked down so it doesn't, anymore.

Was dumb. If we ever lost access to local notifications, we wouldn't be able to provide the mirroring feature at all, so that's probably not a worry. Something like Android is now doing, where it doesn't show the _contents_ of certain (supposedly-sensitive) notifications, might happen, but the notifications themselves aren't likely to go away.

This doesn't mean I really think GSConnect _should be_ trying to do this, but the reason **isn't** because we can't / won't be able to.

@jnnkB 

> If you only sometimes run appB on your laptop and also your laptop and phone aren't always connected, de-duping notifications would ensure that you are notified as often as possible.

I mean... to be pedantic, _NOT_ de-duping notifications will actually ensure you're notified as often as possible. **#JustSayin** 😀 

But more seriously...

> While this is the easiest solution, it's not the most convenient: If you only sometimes run appB on your laptop and also your laptop and phone aren't always connected, de-duping notifications would ensure that you are notified as often as possible.

That's the thing, though. Turning off notification **mirroring** in KDE Connect will _only_ eliminate duplicates. You'll still get the notification on Linux if the app is running. You'll still get the notification on your phone. The only thing you _won't_ get is a possible second notification on Linux, **if** your phone gets a notification _and_ is connected to GSConnect.

The only notifications it can possibly eliminate are duplicates, because they're always mirrored notifications. (That's if we consider the total of notifications across _all_ devices taken together, which I admit is cheating a little.)

1. Is `$APP` running on Linux?
2. Is phone connected to GSConnect?
3. Is notification mirroring enabled for `$APP`?

| 1 | 2 | 3 | Without deduplication | With deduplication |
|---|---|---|-----------------------|--------------------|
|Yes|Yes|Yes|3 notifications (2×Linux+device)|2 notifications (Linux+device)|
|Yes|No |Yes|2 notifications (Linux+device)|2 notifications (Linux+device)|
|No |Yes|Yes|2 notifications (Linux+device)|2 notifications (Linux+device)|
|No |No |Yes|1 notification (device)|1 notification (device)|
|---|---|---|------------------------------|------------------------------|
|Yes|Yes|No |2 notifications (Linux+device)|2 notifications (Linux+device)|
|Yes|No |No |2 notifications (Linux+device)|2 notifications (Linux+device)|
|No |Yes|No |1 notification (device)|1 notification (device)|
|No |No |No |1 notification (device)|1 notification (device)|

If I've gotten all that right (fingers crossed), the **only** one of those scenarios that deduplication even affects is the first one. And the only thing you lose by going the disable-mirroring route, instead, is the second notification that would be mirrored to Linux from the device when it's connected to GSConnect and the `$App` isn't running on Linux. (The No-Yes-Yes vs. No-Yes-No rows.)

That makes deduplication a **lot** of added complexity for fairly minimal benefit, when compared to other solutions.

**Comment by ferdnyc** [comment]
(I already had to edit to fix one row, so anyone following this via email, don't trust or respond to the table in the email notification. It's corrected at https://github.com/GSConnect/gnome-shell-extension-gsconnect/issues/1070#issuecomment-3381294424)

**Comment by ferdnyc** [comment]
Given the above (I realize it's a lot of dense information, apologies):

1. Does anyone see any mistakes I made in the "matrix"? (I double-checked it a couple of times, but I'm not perfect!)
2. If the table is correct, and de-duplication would affect **only** the scenario represented by the first row, and not any of the other seven, does anyone still want to advocate for it as a feature?

Putting together the table above convinced me (if I hadn't been leaning that way already) that a notification deduplication feature doesn't appear to be worth the effort — there's far too little benefit to it, given the significant complexity it would involve.

But I'm open to disagreement. Change my mind! Otherwise, I'd like to close this as unplanned.

**Comment by daniellandau** [comment]
I don't know about the table but IMO this is not a viable feature.

**Comment by stuaxo** [comment]
I'm happy for this to be closed here.   If some 3rd party thing arrives that can find dupes arrives and needs some integration with GSConnect then we could consider re-opening.
